2023-07-28-on-demand-builders.md

```diff
@@ -0,0 +1,125 @@
+On demand started builders in Copr
+==================================
+
+The problem is that our builders are too slow to build e.g. Chromium package
+quickly enough.  Such builds take 24h and more (our limit is 48h).  With the
+cloud era, it really makes sense to have more powerful machines -> even though
+they cost more $$/hour, we spend less amount of time on them.
+
+The problem
+-----------
+
+The thing is that, as of now, every builder in Copr is started by Resalloc "in
+advance", or "preallocated".  The builders are preallocated even though they are
+not _immediately_ used, and stay idle till someone takes them (if there's no
+demand for days, they stay allocated).  When taken, system allocates additional
+resources (so we get more and more machines if there's actual demand for them).
+
+This system works well for normal builders, because they are typically taken
+very quickly (the more tasks in queue the faster they are taken; the smaller the
+"prealloc" quota number for resalloc pool is the faster they are taken).  And
+even if they are not taken for a while, we can actually afford keeping a few of
+them allocated because the instances are not overly expensive.
+
+There's though a demand to have a more (maybe ridiculously) powerful instances;
+having just one of them allocated but unused could be quite expensive matter.
+Even if we used for a low-demanding task.
+
+
+The proposal overview
+---------------------
+
+Let's have on-demand resources/pools in Resalloc.  In such pools there would be
+**no preallocated resource** by default, as long as there's no demand for them.
+Client will have a possibility to ask for them via special "on demand" tagged
```

Comment:
I don't have a proposal for #2364 yet, so I don't know how it will be done yet. But one of the possibilities is an on-demand button for spawning a debugging builder. In such case we would need to make sure the "on demand" tag is general enough to be used for both of these features or maybe use something like beefy/high-performance for this feature to make it more clear.

Reply:
I am opposed to the idea that clients will be the ones to decide whether they want/need high-performance workers. Then we will be forced to write some prevention for misuse.
I would rather keep this decision on our side. In our code. 
For the start I think we can have the logic pretty dumb, e.g., `if name in ['chromium', 'libre-office', ...]` - likely config driven. And later add heuristics. E.g. if your build lasted more than 10 hours then next time you will get high-performance builder.

Reply:
My brain always switches me from the "easy problem/lovely to implement" to "no-go/not implementable" feeling which depresses me :-)  (I'd hate to experience the `research-development => research-only` switch, this is against my nature).

But lemme name some practical problems that we should solve with such a heuristic-only solution:

- events that trigger builds of in multiple projects, having multiple chroots enabled (typically commit to Fedora dist-git)
- time spent on the build isn't really telling much about the build (is it IO/net/memory/...) 
- we waste a lot of storage resources on our builders for tmpfs+SWAP (an extreme example is that we have x86 machines >= 400G volumes), I kind of believe this is solvable by the same "I want more swap" feature.
- consider that most of the builds can be done in-container;  I thought this is again solvable by the same feature like "give me container builders"
- there are teams that are able to pay some money (give us a cloud budget) to get some boost for themselves, and again I thought this is solvable by the same feature

> Then we will be forced to write some prevention for misuse.

What I'm really afraid of is the price we'll pay and the gain we can get for it, and whether we can really "sell it" as a benefit (rather than having PR-only thing related to AI). Right now we have existing requests from highly experienced and trusted users that we can help to easily. We also trust the Copr user-base in general (we are not generally afraid of misuse, or face attacks on a daily basis).

I'm fine with going out with some simple heuristic (both adding or removing the need for a high-performance builder), but not having a way to ask explicitly seems at least like a lost opportunity.

Reply:
What's also worth noting is that asking for high-performance builder is not "for free" to blindly over-use it; users will have to wait till the machine is started in the first place, and because of much lower limits, they will very likely stay longer in the queue (is it better to stay hour or two in the queue, and then save several hours at build time?).